### PR TITLE
feat: add Koa2 support

### LINKED ISSUE #8: Investigate Supporting koa@next
###### Copied from original issue: https://github.com/GoogleCloudPlatform/google-cloud-node/issues/2295

<a href="/DominicKramer"></a><a href="/DominicKramer">@&shy;DominicKramer</a><br>May 11, 2017 9:37 PM

The cloud-errors-nodejs repository has PR [#97](https://github.com/GoogleCloudPlatform/cloud-errors-nodejs/pull/97) that already contains some preliminary work.

### PR DESCRIPTION
**Pull request by mike-marcacci**
Fixes #8 

- [x] Tests and linter pass<sup>1</sup>
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)<sup>2</sup>

<sup>1</sup> I added tests to "system-tests", which I didn't get running locally. Also, the existing koa test *should* be failing, given that the code targets koa@1 while it depends on koa@2. I assume these aren't used?

<sup>2</sup> Docs for all middleware are missing from `README.md`, so I didn't add any, and `README_OLD.md` is clearly "old".

### CODE REVIEW

**Comment by mike-marcacci** [comment]
CLA signed.

**Comment by DominicKramer** [comment]
Awesome.  Thanks for contributing this.  I think it looks good, but I need to think more about whether there should be separate `koa` and `koa2` properties.  For `hapi`, the `hapi` property works for both `hapi6` and `hapi7`.  I'm not sure if it possible to do this with `koa`.

**Comment by mike-marcacci** [comment]
Hey @DominicKramer - I changed this to work with both koa1 and koa2. I did escape the type system a bit with some `any` casts, which probably isn't ideal. Feel free to add better typing, or perhaps I'll have time to look at that soon.

**Comment by DominicKramer** [comment]
@mike-marcacci Thanks for working on this.  It looks really good.  I will take care of updating the types for you.  Thanks.

**Comment by mike-marcacci** [comment]
Awesome! Thanks so much @DominicKramer. I did also want to reiterate my note from the original PR: I'm pretty sure the "system-tests" aren't actually being run, at least locally with `npm test`. You might want to confirm this before merging incoming PRs in general :)

**Comment by DominicKramer** [comment]
@mike-marcacci Sorry, I missed your "system tests" comment. It is expected that the system tests are not run with `npm test`.  The system tests and samples tests are run separately using the `npm run system-test` and `npm run samples-test` commands and are handled by CircleCI.  Thank you, though, for asking about this.

**Comment by mike-marcacci** [comment]
🤦🏻‍♂️ I guess I should have read the package.json. Looks like I inadvertently pulled in some greenkeeper commits and broke the CLA googlebot, so I rebased and forced-pushed.

**Comment by DominicKramer** [comment]
No worries.

**Comment by DominicKramer** [comment]
There were a few issues unrelated to this PR that are blocking this PR from landing.  Some are fixed in PR  #183 and PR #185.  There is another in the pipeline.  When those are done, this PR should be unblocked.

File: src/interfaces/koa.ts [hunk 1]
```diff
@@ -58,4 +58,4 @@ export function koaErrorHandler(client: RequestHandler, config: Configuration) {
       client.sendError(em);
     }
   };
-}
+}
```
Comment (kjin) [inline src/interfaces/koa.ts]: Can you preserve the newline here?

File: src/interfaces/koa2.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright 2016 Google Inc. All Rights Reserved.
```
Comment (kjin) [inline src/interfaces/koa2.ts]: Should this be `2018` since it's a new file?

Comment (DominicKramer) [reply src/interfaces/koa2.ts]: Yep good catch.

File: src/interfaces/koa.ts (hunk 1 above)
Comment (DominicKramer) [reply src/interfaces/koa.ts]: Will do.

**Pull request merged**